icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,462 @@
+// © 2022 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#include <utility>
+
+#include "unicode/utypes.h"
+
+#if !UCONFIG_NO_BREAK_ITERATION
+
+#include "cjmlbe.h"
+#include "brkeng.h"
+#include "dictbe.h"
+#include "unicode/uniset.h"
+#include "unicode/chariter.h"
+#include "unicode/resbund.h"
+#include "unicode/ubrk.h"
+#include "unicode/usetiter.h"
+#include "ubrkimpl.h"
+#include "utracimp.h"
+#include "uvectr32.h"
+#include "uvector.h"
+#include "uassert.h"
+#include "unicode/normlzr.h"
+#include "cmemory.h"
+#include "dictionarydata.h"
+
+U_NAMESPACE_BEGIN
+
+#if !UCONFIG_NO_NORMALIZATION
+
+class Element : public UMemory {
+public:
+    /**
+     * Default constructor.
+     *
+     * @param ch A Unicode character
+     * @param index The unicode block index of the character.
+     */
+    Element(UChar32 ch, UnicodeString index) {
+        this->ch = ch;
+        this->index = index;
+    }
+
+    UChar32 ch;
+    UnicodeString index;
+};
+
+CjMLBreakEngine::CjMLBreakEngine(UnicodeSet& fDigitOrOpenPunctuationOrAlphabetSet,
+    UnicodeSet& fClosePunctuationSet, UErrorCode& status) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+    this->fDigitOrOpenPunctuationOrAlphabetSet = fDigitOrOpenPunctuationOrAlphabetSet;
+    this->fClosePunctuationSet = fClosePunctuationSet;
+    fUnicodeBlock = new UVector32(status);
+    loadMLparams(status);
+}
+
+CjMLBreakEngine::~CjMLBreakEngine() {
+    fUnicodeBlock->removeAllElements();
+    delete fUnicodeBlock;
+    fModel.removeAll();
+    fDigitOrOpenPunctuationOrAlphabetSet.removeAllStrings();
+    fClosePunctuationSet.removeAllStrings();
+}
+
+static const UChar32 INVALID = '|';
+
+static UnicodeString* concatChar(const char* str, UChar32 arr[], UErrorCode& status) {
+    UnicodeString* temp = new UnicodeString(str);
```

Comment:
This constructor invokes the default-charset converter. We should never use that in ICU library code. We should know what encoding it is, and call an appropriate constructor. If it's just "invariant characters" (subset of ASCII), then use `UnicodeString(str, -1, US_INV)`.

Reply:
Done. Thanks for the suggestion.